src/ccs/mod.rs

```diff
@@ -0,0 +1,509 @@
+/// We also have t, q, d constants and c (vector), S (set)
+/// As well as m, n, s, s_prime
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCSShape<G: Group> {
+  pub(crate) M: Vec<SparseMatrix<G::Scalar>>,
+  // Num vars
+  pub(crate) t: usize,
+  // Number of public witness
+  pub(crate) l: usize,
+  pub(crate) q: usize,
+  pub(crate) d: usize,
+  pub(crate) S: Vec<Vec<usize>>,
+
+  // Was: usize
+  pub(crate) c: Vec<G::Scalar>,
+
+  // n is the number of columns in M_i
+  pub(crate) n: usize,
+  // m is the number of rows in M_i
+  pub(crate) m: usize,
+  // s = log m
+  pub(crate) s: usize,
+  // s_prime = log n
+  pub(crate) s_prime: usize,
+}
+
+impl<G: Group> CCSShape<G> {
+  /// Create an object of type `CCSShape` from the explicitly specified CCS matrices
+  pub fn new(
+    M: &[SparseMatrix<G::Scalar>],
+    t: usize,
+    l: usize,
+    q: usize,
+    d: usize,
+    S: Vec<Vec<usize>>,
+    c: Vec<G::Scalar>,
+  ) -> CCSShape<G> {
+    // Can probably be made more efficient by keeping track fo n_rows/n_cols at creation/insert time
+    let m = M
+      .iter()
+      .fold(usize::MIN, |acc, matrix| max(acc, matrix.n_rows()));
+    let n = M
+      .iter()
+      .fold(usize::MIN, |acc, matrix| max(acc, matrix.n_cols()));
+
+    // Check that the row and column indexes are within the range of the number of constraints and variables
+    assert!(M
+      .iter()
+      .map(|matrix| matrix.is_valid(m, t, l))
+      .collect::<Result<Vec<()>, NovaError>>()
+      .is_ok());
+
+    // We require the number of public inputs/outputs to be even
+    assert_ne!(l % 2, 0, " number of public i/o has to be even");
+
+    let s = m.log_2();
+    let s_prime = n.log_2();
+
+    CCSShape {
+      M: M.to_vec(),
+      t,
+      l,
+      q,
+      d,
+      S,
+      c,
+      m,
+      n,
+      s,
+      s_prime,
+    }
+  }
+
+  pub(crate) fn to_cccs_shape(&self) -> CCCSShape<G> {
+    let M_mle = self.M.iter().map(|matrix| matrix.to_mle()).collect();
+    CCCSShape {
+      M_MLE: M_mle,
+      ccs: self.clone(),
+    }
+  }
+
+  // Transform the CCS instance into a CCCS instance by providing a commitment key.
+  pub fn to_cccs<R: RngCore>(
+    &self,
+    rng: &mut R,
+    ck: &<<G as Group>::CE as CommitmentEngineTrait<G>>::CommitmentKey,
+    z: &[G::Scalar],
+  ) -> (CCCSInstance<G>, CCSWitness<G>, CCCSShape<G>) {
+    let w: Vec<G::Scalar> = z[(1 + self.l)..].to_vec();
+    // XXX: API doesn't offer a way to handle this apparently? Need to investigate
+    let _r_w = G::Scalar::random(rng);
+    let C = <<G as Group>::CE as CommitmentEngineTrait<G>>::commit(ck, &w);
+
+    (
+      CCCSInstance {
+        C,
+        x: z[1..(1 + self.l)].to_vec(),
+      },
+      CCSWitness { w },
+      self.to_cccs_shape(),
+    )
+  }
+
+  /// Transform the CCS instance into an LCCCS instance by providing a commitment key.
+  pub fn to_lcccs<R: RngCore>(
```

Comment:
I'm a bit confused about this code. In particular, I'm not sure what it means to transform a CCS instance into an LCCS instance---since this seems implausible as LCCCS can only express linear constraints (a la addition gates).